src/components/Search/Results.tsx

```diff
@@ -1,25 +1,83 @@
 import { Link } from "@/components/Link";
 import { Search } from "@/types";
 import { Markup } from "interweave";
-import React from "react";
+import { useRouter } from "next/router";
+import React, { useCallback, useEffect, useState } from "react";
+import tinykeys from "tinykeys";
 import tw from "twin.macro";
 
 interface Props {
-  closeModal: () => void
+  closeModal: () => void;
   results: Search.Result;
 }
 
+type SelectedResult = { idx: number; slug: string };
+
+const withoutBaseUri = (slug: string) => {
+  const url = new URL(slug);
+  const { hash, pathname } = url;
+  return `${pathname}${hash}`;
+};
+
+// @FIXME: Indexer is grabbing #__next from anchor hrefs. This should be
+// fixed upstream, but no harm in just hacking it in place for now.
+const cleanSlug = (slug: string) => withoutBaseUri(slug.replace("#__next", ""));
+
 const Results: React.FC<Props> = ({ closeModal, results }) => {
-  const withoutBaseUri = (slug: string) => {
-    const url = new URL(slug);
-    const { hash, pathname } = url;
-    return `${pathname}${hash}`;
-  };
+  const router = useRouter();
+
+  // A result is selected when navigated to using arrow keys, or on mouse
+  // hover.
+  const [selectedResult, setSelectedResult] = useState<SelectedResult | null>(
+    null,
+  );
+
+  const resultsFlat: SelectedResult[] = Object.values(results)
+    .flat()
+    .map((r, idx) => ({ idx, slug: cleanSlug(r.slug) }));
 
-  // @FIXME: Indexer is grabbing #__next from anchor hrefs. This should be
-  // fixed upstream, but no harm in just hacking it in place for now.
-  const cleanSlug = (slug: string) =>
-    withoutBaseUri(slug.replace("#__next", ""));
+  const onArrowKeyDown = useCallback(() => {
+    if (selectedResult && selectedResult.idx + 1 >= resultsFlat.length) {
+      // End of results; nothing to go down from.
+      return;
+    }
+
+    setSelectedResult(prev => {
+      // On key down, go to the next item.
+      const next = prev ? resultsFlat[prev.idx + 1] : resultsFlat[0];
+      return { ...next };
+    });
+  }, [resultsFlat, selectedResult, setSelectedResult]);
+
+  const onArrowKeyUp = useCallback(() => {
+    setSelectedResult(prev => {
+      if (prev === null || prev.idx === 0) {
+        // Start of results. Going up from here is the search input, so it's
+        // a no-op.
+        return null;
+      }
```

Comment:
This would ideally also tweak the input focus:
* on key down, remove focus on input
* on key up to here (idx=0/head of result), re-focus input

Requires some `ref` trickery to wire up. I don't want to get into that right now. Also may get somewhat confusing since it requires different stateful props to be passed around (between Modal, QueryInput, Results I think), so it's best to do that in a separate pull and keep this lean.

Reply:
I agree here, best to have the listener like this until we can do something more ideal. Good call.